### ISSUE TITLE
[Feature] Add an option to pivot a table view (Table Visualization)

### ISSUE THREAD

**Issue by Totalus** [grafana/grafana#76089]
I would like to turn this:

![image](https://github.com/grafana/grafana/assets/34516402/8e85d7d0-02a8-4bab-8ed3-a27ca80804f2)

Into this:

![image](https://github.com/grafana/grafana/assets/34516402/eb8fd93a-61d9-4333-8448-9f5c1f85233d)

Which is more space efficient and easier to read.

The idea would be to add an option in the _Table_ visualization that allows to simply pivot the table without changing the data structure (the data frames would remain the same but simply become displayed horizontally as rows instead of as columns). It is purely visual and would allow to better view tables that have a lot of columns and a low number of rows.

I would see it as a simple option of the view:

![image](https://github.com/grafana/grafana/assets/34516402/6c64c88a-9ac1-4953-96e5-cb8e4e097d16)

**Comment by venkatbvc** [comment]
i will contribute this feature.

**Comment by khushijain21** [comment]
Can you direct me to the file names to get the Pivot Button started?

**Comment by Totalus** [comment]
Probably [Table.ts](https://github.com/grafana/grafana/blob/main/packages/grafana-ui/src/components/Table/Table.tsx) and some other files around that.

**Comment by khushijain21** [comment]
@Totalus should PivotTable when enabled by horizontal or vertical? And what should be the default option?

**Comment by Totalus** [comment]
I would say the PivotTable option should be disabled by default and the table should display normally (as it currently is without the option).
When PivotTable is enabled, then it would pivot rows and columns. That is what I would expect for this features.

**Comment by khushijain21** [comment]
Hello @Totalus, I was trying this feature, but as I see in this issue https://github.com/TanStack/table/issues/49 that pivoting table is not supported by react-table just yet. How were you able to transform the table by code in the picture you attached above? Could you share that understanding?

**Comment by Totalus** [comment]
> How were you able to transform the table by code in the picture you attached above? 

I just took a screen capture of the original table layout and used a drawing tool (paint-like) to arrange them vertically to illustrate the idea.

As for how to exactly implement the feature, I did not really dig into the code to see how it could be done.

> [...] pivoting table is not supported by react-table just yet.

Reshaping the data provided to the table could possibly work, as https://github.com/TanStack/table/issues/49#issuecomment-396429133 did (be aware that though it could be useful, this issue belong to a table library that does not seem to be used by grafana). Basically if you can shape the data so that the rows become the columns before passing it to the component that renders it, it might just work.

**Comment by khushijain21** [comment]
> this issue belong to a table library that does not seem to be used by grafana

I see we use react-table right? This issue is from the same repo. 

Sorry for so many questions but I see two approaches here:
1. If we reshape data 
    - The header will have to be positioned to absolute on the left - for horizontal scrolling
    - The footer Options will have to be moved too
    - The logic must also incorporate nested data fields and then we can transpose our json data
    
 2. Other approach
    - Do we contribute to react-table to get this feature out of the box?

**Comment by Totalus** [comment]
> I see we use react-table right? This issue is from the same repo.

You're right, I didn't realize the latest version (v8) of React-Table had been renamed TanStack Table.

As for what would be the best approach, the short answer is _I don't know_, but here are a few comments I can make:

- The use case I can see for pivoting the table is when you have a small number of rows and high number of columns that you all want to display. In this use case I don't think the footer option is really useful, so we could probably just not render the footer when the table is pivoted, unless (or until) someone finds a useful use case for the footer on a pivoted table.

- I haven't personally been using the nested data feature, but we could possibly apply the same logic as for the footer.

- If the pivot feature is implemented within react-table, it will probably take much more time to be integrated into grafana. This is because it must first be integrated in a release of react-table, then the library needs to be updated in grafana, then we can actually implement the feature. Currently grafana is using react-table v7 and from a quick search it is not clear if there is plan to upgrade to v8 soon (and v8 brings breaking changes that requires a certain amount of migration, so its not like just updating a minor or patch version, it may take more time to happen). For the feature to be integrated into grafana in a time-reasonable frame, it would probably have to be added to react-table v7 (that is if they still accept new features in v7). It should also be added to v8, because if not, the day grafana upgrades to react-table v8, the pivot feature won't work anymore.

Reshaping the data is one idea, but don't close your mind to other options. My approach to this would be to read and play around with the code related to the Table panel, until I understand the architecture and then think and find one (or more) places where the pivot could be made, ideally without too much effort. Then test it out and see what result it gives, probably iterate on that to fix the issues that will come along, or find another method if the first option turns out to have too many limitations or issues and an alternative might work better.

**Comment by mdvictor** [comment]
Hey @Totalus. Thank you for opening this issue!

I think the implementation of a Pivot Table would be quite an undertake. We are currently in discussions about such a feature but there is not ETA on when/if it will get released. For this specific scenario I would say a transformation reshaping the data, similar to 'Rows to fields', would achieve what's needed. Or maybe a change in `Rows to fields` to achieve the inverse and also do fields to rows. In any case, a transformation approach would be my take on it. We will further look into this issue and prioritise accordingly.

**Comment by justinsmalley** [comment]
There is a "Transpose" transform available now that I have started using for this purpose.

**Comment by costela** [comment]
Just FYI: the "Transpose" transform kinda works, but has at least one considerable limitation: it's not possible to add overrides per field, since there's just two fields after the transform. This means it's not possible to - e.g. - add data links or set different formats to different transposed fields.
